src/Measurement_Observation/haskell_model.rst

```diff
@@ -0,0 +1,129 @@
+
+Haskell Compilation and Execution Model
+=======================================
+
+
+GHC doesn't use a VM but compiles to native code (when a native code generator
+is used).
+So we can use tools to profile native code execution: e.g. perf.
+However the native code we execute doesn't look exactly like the one produced by
+imperative languages (C, C++, Rust, etc.) and this confuses the tools which make
+some assumptions.
+For example:
+
+1. we don't have clear procedure boundaries with call/return instruction pairs
+   (we use tail calls, i.e. jump instructions)
+
+2. we don't use the so-called C stack in the usual way. I.e. we don't use usual
+   stack registers (e.g. rsp/rbp on x86-64) as stack registers.
+
+3. GHC uses its own calling convention
+
+In addition, lazy evaluation makes control-flow and memory usage difficult to
+understand.
+
+In this chapter we describe the compilation pipeline and the execution model.
+For each stage of the pipeline we list what can be measured at runtime.
+
+
+Compilation pipeline overview
+-----------------------------
+
+GHC compiles Haskell codes in several stages, each with their own intermediate
+representation:
+
+Haskell: the functional language we love with its bazillion of extensions.
+
+Core: a simple and explicitly typed functional language. Types are basically
+Haskell types. Type-class dictionaries are passed explicitly as records,
+type-applications are used everywhere needed, coercions (proofs that we can
+convert a type into another) are first class values, etc.
```

Comment:
```suggestion
convert a type into another) are first class values, etc.

Core is a good intermediate representation to observe the effect of polymorphism and optimization passes on your code. Core should also be used as a first stop on the journey into the pipeline because it is the intermediate representation that will be closest to the familiar Haskell code.
```